slsa/provenance.go

```diff
@@ -0,0 +1,123 @@
+const GithubHostedActionsBuilderID = "https://github.com/Attestations/GitHubHostedActions@v1"
+
+// WorkflowRun contains information about the build run including the builder,
+// build invocation, and parameters.
+type WorkflowRun struct {
+	// Subjects is information about the generated artifacts.
+	Subjects []intoto.Subject
+
+	// BuildType indicates the type of build that was done. More importantly it
+	// also specifies the format of the BuildConfig.
+	BuildType string
+
+	// BuildConfig is metadata about the build.
+	BuildConfig interface{}
+
+	// GithubContext is the context for the workflow run.
+	GithubContext github.WorkflowContext
+}
+
+var (
+	parametersVersion = 1
+)
+
+// WorkflowParametersV1 contains parameters given to the workflow invocation in v1 format.
+type WorkflowParametersV1 struct {
+	// Version is the version of the
+	Version int `json:"version"`
+
+	// EventName is the name of the event that initiated the workflow run.
+	EventName string `json:"event_name,omitempty"`
+
+	// EventPayload is the full event payload.
+	// See: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows
+	EventPayload interface{} `json:"event_payload,omitempty"`
+
+	// RefType is type of ref that triggered the workflow run.
+	RefType string `json:"ref_type,omitempty"`
+
+	// Ref is the ref that triggered the workflow run.
+	Ref string `json:"ref,omitempty"`
+
+	// BaseRef is the base ref or base branch of the pull request in a workflow
+	// run.
+	BaseRef string `json:"base_ref,omitempty"`
+
+	// HeadRef is ref or source branch of the pull request in a workflow run.
+	HeadRef string `json:"head_ref,omitempty"`
+
+	// Actor is the username of the user that initiated the workflow run.
+	Actor string `json:"actor,omitempty"`
+
+	// SHA1 is the commit SHA that triggered the workflow run.
+	SHA1 string `json:"sha1,omitempty"`
+}
+
+// HostedActionsProvenance generates an in-toto provenance statement in the SLSA
+// v0.2 format for a workflow run on a Github actions hosted runner.
+func HostedActionsProvenance(w WorkflowRun) (intoto.ProvenanceStatement, error) {
+	return intoto.ProvenanceStatement{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: slsa.PredicateSLSAProvenance,
+			Subject:       w.Subjects,
+		},
+		Predicate: slsa.ProvenancePredicate{
+			BuildType: w.BuildType,
+			Builder: slsa.ProvenanceBuilder{
+				ID: GithubHostedActionsBuilderID,
+			},
+			Invocation: slsa.ProvenanceInvocation{
+				ConfigSource: slsa.ConfigSource{
+					EntryPoint: w.GithubContext.Workflow,
+					URI:        w.GithubContext.RepositoryURI(),
+					Digest: slsa.DigestSet{
+						"sha1": w.GithubContext.SHA,
+					},
+				},
+				// Non user-controllable environment vars needed to reproduce the build.
+				Environment: map[string]interface{}{
+					// NOTE: Hosted runners are always amd64.
+					"arch":               runtime.GOARCH,
```

Comment:
nit: I would omit this since it's misleading - it tells us the architecture of the provenance generator, which may differ from the architecture that the actual build job took place on.

Reply:
I don't think that will address mark's comment which is that the calling job could technically be a different architecture than the job running the provenance generation.

The only way I see that we can do that is to add it as an input.

I added issue #5 and we can have further conversation there. I removed the "arch" key for now.

Reply:
SG, thanks.